File: stepik-union/src/main/java/org/stepik/core/courseFormat/Node.java
package org.stepik.core.courseFormat;

import com.intellij.ide.projectView.ProjectView;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.thoughtworks.xstream.annotations.XStreamOmitField;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.exceptions.StepikClientException;
import org.stepik.api.objects.StudyObject;
import org.stepik.api.objects.progresses.Progresses;
import org.stepik.core.stepik.StepikConnectorLogin;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * @author meanmail
 */
abstract class Node<
        D extends StudyObject,
        C extends StudyNode<DC, CC>,
        DC extends StudyObject,
        CC extends Node> implements StudyNode<D, C> {
    private static final Logger logger = Logger.getInstance(Node.class);
    private static final ExecutorService executor = Executors.newSingleThreadExecutor();
    private StudyNode parent;
    private D data;
    private List<C> children;
    private boolean wasDeleted;
    @XStreamOmitField
    private volatile StudyStatus status;
    @XStreamOmitField
    private Project project;

    Node() {
    }

    Node(@NotNull Project project, @NotNull D data) {
        setData(data);
        init(project, null);
    }

    @Nullable
    public Project getProject() {
        return project;
    }

    public void setProject(@NotNull Project project) {
        if (this.project == project) {
            return;
        }
        this.project = project;
        for (C child : getChildren()) {
            child.setProject(project);
        }
    }

    @Nullable
    private StudyNode getLastNode() {
        List<C> children = getChildren();

        for (int i = children.size() - 1; i >= 0; i--) {
            C child = children.get(i);
            if (!child.getWasDeleted()) {
                return child;
            }
        }
        return null;
    }

    @Nullable
    private StudyNode getFirstNode() {
        for (C child : getChildren()) {
            if (!child.getWasDeleted()) {
                return child;
            }
        }

        return null;
    }

    @Nullable
    @Override
    public StudyNode getPrevChild(@Nullable StudyNode current) {
        List<C> children = getChildren();
        if (current == null) {
            return getLastNode();
        }

        int position = current.getPosition();

        for (int i = children.size() - 1; i >= 0; i--) {
            StudyNode item = children.get(i);
            if (!item.getWasDeleted() && item.getPosition() < position) {
                return item;
            }
        }
        return null;
    }

    @Nullable
    @Override
    public StudyNode getNextChild(@Nullable StudyNode current) {
        List<C> children = getChildren();
        if (current == null) {
            return getFirstNode();
        }

        int position = current.getPosition();
        for (StudyNode item : children) {
            if (!item.getWasDeleted() && item.getPosition() > position) {
                return item;
            }
        }
        return null;
    }

    @Override
    public boolean isLeaf() {
        return getChildren().isEmpty();
    }

    @Nullable
    @Override
    public StudyNode getParent() {
        return parent;
    }

    public void setParent(@Nullable StudyNode parent) {
        this.parent = parent;
    }

    @NotNull
    @Override
    public String getPath() {
        if (parent != null) {
            if (parent.getPath().isEmpty()) {
                return getDirectory();
            }
            return parent.getPath() + "/" + getDirectory();
        } else {
            return "";
        }
    }

    @NotNull
    @Override
    public String getDirectory() {
        if (parent != null) {
            return getDirectoryPrefix() + getId();
        } else {
            return "";
        }
    }

    @NotNull
    String getDirectoryPrefix() {
        return "";
    }

    @Nullable
    @Override
    public C getChildById(long id) {
        for (C child : getChildren()) {
            if (child.getId() == id) {
                return child;
            }
        }

        return null;
    }

    @Nullable
    @Override
    public StudyNode<?, ?> getChildByClassAndId(@NotNull Class<? extends StudyObject> clazz, long id) {
        if (getChildDataClass() == clazz) {
            for (C child : getChildren()) {
                if (child.getId() == id) {
                    return child;
                }
            }
        } else {
            for (C child : getChildren()) {
                StudyNode<?, ?> node = child.getChildByClassAndId(clazz, id);
                if (node != null) {
                    return node;
                }
            }
        }
        return null;
    }

    @Nullable
    @Override
    public C getChildByPosition(int position) {
        for (C child : getChildren()) {
            if (child.getPosition() == position) {
                return child;
            }
        }

        return null;
    }

    @NotNull
    private Map<Long, C> getMapNodes() {
        HashMap<Long, C> mapNodes = new HashMap<>();
        getChildren().forEach(node -> mapNodes.put(node.getId(), node));
        return mapNodes;
    }

    private void sortChildren() {
        getChildren().sort(StudyNodeComparator.getInstance());
    }

    @Override
    public List<C> getChildren() {
        if (children == null) {
            children = new ArrayList<>();
        }

        return children;
    }

    protected abstract List<DC> getChildDataList();

    @Override
    public void init(@NotNull Project project, @Nullable StudyNode parent) {
        this.project = project;
        setParent(parent);
        Map<Long, C> mapNodes = getMapNodes();
        List<C> processed = new ArrayList<>();

        for (DC data : getChildDataList()) {
            C child = mapNodes.get(data.getId());
            if (child == null) {
                try {
                    child = getChildClass().newInstance();
                    getChildren().add(child);
                } catch (InstantiationException | IllegalAccessException e) {
                    logger.warn("Can't get new instance for child", e);
                    break;
                }
            }
            child.setData(data);
            child.setWasDeleted(wasDeleted);
            processed.add(child);
        }

        sortChildren();
        ArrayList<C> wasDeletedList = new ArrayList<>(getChildren());
        wasDeletedList.removeAll(processed);
        wasDeletedList.forEach(child -> child.setWasDeleted(true));

        for (StudyNode child : getChildren()) {
            child.init(project, this);
        }
    }

    @Override
    public void reloadData(@NotNull Project project) {
        if (loadData(data.getId())) {
            init(project);
        } else {
            setProject(project);
        }
    }

    protected abstract boolean loadData(long id);

    protected abstract Class<C> getChildClass();

    protected abstract Class<DC> getChildDataClass();

    @Override
    public boolean isUnknownStatus() {
        return status == null;
    }

    @NotNull
    @Override
    public StudyStatus getStatus() {
        if (isUnknownStatus()) {
            status = StudyStatus.UNCHECKED;

            executor.execute(() -> {
                try {
                    Map<String, StudyNode> progressMap = new HashMap<>();
                    Node.this.getChildren().stream()
                            .filter(StudyNode::isUnknownStatus)
                            .forEach(child -> {
                                DC data = child.getData();
                                if (data != null) {
                                    progressMap.put(data.getProgress(), child);
                                }
                            });
                    D data = Node.this.getData();
                    if (data != null) {
                        String progressId = data.getProgress();
                        if (progressId != null) {
                            progressMap.put(progressId, Node.this);
                        }

                        Set<String> progressIds = progressMap.keySet();

                        if (!progressIds.isEmpty()) {
                            int size = progressIds.size();
                            String[] list = progressIds.toArray(new String[size]);
                            int start = 0;
                            int end;
                            while (start < size) {
                                end = start + 20;
                                if (end > size) {
                                    end = size;
                                }
                                String[] part = Arrays.copyOfRange(list, start, end);
                                start = end;

                                StepikApiClient stepikApiClient = StepikConnectorLogin.authAndGetStepikApiClient();

                                Progresses progresses = stepikApiClient.progresses()
                                        .get()
                                        .id(part)
                                        .execute();

                                progresses.getItems().forEach(progress -> {
                                    String id = progress.getId();
                                    StudyNode node = progressMap.get(id);
                                    if (progress.isPassed()) {
                                        node.setRawStatus(StudyStatus.SOLVED);
                                    }
                                });
                            }
                        }
                    }
                    if (!project.isDisposed()) {
                        ProjectView.getInstance(project).refresh();
                    }
                } catch (StepikClientException e) {
                    logger.warn(e);
                }
            });
        }

        return status;
    }

    @Override
    public void setStatus(@Nullable StudyStatus status) {
        if (status == StudyStatus.SOLVED && this.status != status) {
            this.status = StudyStatus.SOLVED;

            StudyNode parent = getParent();

            while (parent != null) {
                parent.setRawStatus(null);
                parent = parent.getParent();
            }
        }
    }

    @Override
    public void setRawStatus(@Nullable StudyStatus status) {
        this.status = status;
    }

    @Override
    public void passed() {
        setStatus(StudyStatus.SOLVED);
    }

    @Nullable
    @Override
    public D getData() {
        if (data == null) {
            try {
                data = getDataClass().newInstance();
            } catch (InstantiationException | IllegalAccessException e) {
                logger.warn("Can't create data instance: " + getDataClass().getName());
                return null;
            }
        }
        return data;
    }

    @Override
    public void setData(@Nullable D data) {
        this.data = data;
    }

    protected abstract Class<D> getDataClass();

    @Override
    public long getId() {
        StudyObject data = getData();
        return data != null ? data.getId() : 0;
    }

    public void setId(long id) {
        StudyObject data = getData();
        if (data != null) {
            data.setId(id);
        }
    }

    @Override
    public boolean canBeLeaf() {
        return false;
    }

    @Override
    public int getPosition() {
        StudyObject data = getData();
        return data != null ? data.getPosition() : 0;
    }

    @NotNull
    @Override
    public String getName() {
        StudyObject data = getData();
        return data != null ? data.getTitle() : "";
    }

    @Override
    public boolean getWasDeleted() {
        return wasDeleted;
    }

    @Override
    public void setWasDeleted(boolean wasDeleted) {
        this.wasDeleted = wasDeleted;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Node<?, ?, ?, ?> node = (Node<?, ?, ?, ?>) o;

        if (wasDeleted != node.wasDeleted) return false;
        if (data != null ? !data.equals(node.data) : node.data != null) return false;
        //noinspection SimplifiableIfStatement
        if (children != null ? !children.equals(node.children) : node.children != null) return false;
        return status == node.status;
    }

    @Override
    public int hashCode() {
        int result = (data != null ? data.hashCode() : 0);
        result = 31 * result + (children != null ? children.hashCode() : 0);
        result = 31 * result + (wasDeleted ? 1 : 0);
        result = 31 * result + (status != null ? status.hashCode() : 0);
        return result;
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/courseFormat/Node.kt
package org.stepik.core.courseFormat

import com.intellij.openapi.project.Project
import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamOmitField
import org.stepik.api.client.StepikApiClient
import org.stepik.api.exceptions.StepikClientException
import org.stepik.api.objects.StudyObject
import org.stepik.api.objects.progresses.Progresses
import org.stepik.core.auth.StepikAuthManager
import org.stepik.core.auth.StepikAuthManager.isAuthenticated
import org.stepik.core.courseFormat.StudyStatus.FAILED
import org.stepik.core.courseFormat.StudyStatus.SOLVED
import org.stepik.core.courseFormat.StudyStatus.UNCHECKED
import org.stepik.core.utils.batch
import org.stepik.core.utils.refreshProjectView
import java.util.concurrent.Executors

abstract class Node(project: Project? = null,
                    stepikApiClient: StepikApiClient? = null,
                    data: StudyObject? = null) : StudyNode {
    
    override var parent: StudyNode? = null
    
    @XStreamAlias("data")
    private var _data: StudyObject? = null
    
    final override var data: StudyObject
        get() {
            if (_data == null) {
                try {
                    _data = dataClass.newInstance()
                } catch (e: Exception) {
                    logger.warn("Can't create data instance: ${dataClass.name}")
                }
            }
            return _data ?: throw AssertionError("Set to null by another thread")
        }
        set(value) {
            _data = value
        }
    
    @XStreamOmitField
    override var project: Project? = null
    
    init {
        this._data = data
        init(project, stepikApiClient, null)
    }
    
    @XStreamAlias("children")
    private var _children: List<Node>? = null
    
    override val children: List<Node>
        get() {
            if (_children == null) {
                return emptyList()
            }
            return _children ?: throw AssertionError("Set to null by another thread")
        }
    
    fun setChildren(children: List<Node>) {
        _children = children
    }
    
    override var wasDeleted: Boolean = false
    
    @XStreamOmitField
    @Volatile
    private var checkingStatus = false
    
    @XStreamOmitField
    @Volatile
    private var _status: StudyStatus? = UNCHECKED
    
    private fun updateStatus() {
        val stepikApiClient = StepikAuthManager.authAndGetStepikApiClient()
        if (!isAuthenticated) {
            return
        }
        
        val progressIdToNode = children.filter { it._status == UNCHECKED && !it.checkingStatus }
                .associateBy { it.data.progress }
        
        if (progressIdToNode.isNotEmpty()) {
            progressIdToNode.keys.batch(100)
                    .forEach { ids ->
                        var page = 1
                        var progresses: Progresses
                        do {
                            val query = stepikApiClient.progresses()
                                    .get()
                                    .id(*ids.toTypedArray())
                                    .page(page++)
                            try {
                                progresses = query.execute()
                                progresses.forEach { progress ->
                                    val id = progress.id
                                    val node = progressIdToNode[id] ?: if (id == data.progress) this else null
                                    node?.apply {
                                        _status = if (progress.isPassed) SOLVED else FAILED
                                        checkingStatus = false
                                    }
                                }
                            } catch (e: StepikClientException) {
                                logger.warn(e)
                                break
                            }
                        } while (progresses.meta.hasNext)
                    }
        }
        
        refreshProjectView(project)
    }
    
    override var status: StudyStatus
        get() {
            val status = _status
            if ((status == UNCHECKED || status == null) && !checkingStatus) {
                checkingStatus = true
                
                executor.execute {
                    updateStatus()
                }
            }
            
            return status ?: UNCHECKED
        }
        set(value) {
            if (_status !== value) {
                if (value === FAILED && _status === SOLVED) {
                    return
                }
                _status = value
                var parent = parent
                while (parent != null) {
                    parent.status = UNCHECKED
                    parent = parent.parent
                }
            }
        }
    
    override val isLeaf: Boolean
        get() = children.isEmpty()
    
    override val path: String
        get() {
            val parentPath = parent?.path ?: return ""
            
            if (parentPath.isEmpty()) {
                return directory
            }
            
            return "$parentPath/$directory"
        }
    
    override val directory: String
        get() {
            parent ?: return ""
            return "$directoryPrefix$id"
        }
    
    open val directoryPrefix
        get() = ""
    
    private val mapNodes: Map<Long, StudyNode>
        get() = children.associateBy { it.id }
    
    protected abstract val childClass: Class<out Node>
    
    protected abstract val childDataClass: Class<out StudyObject>
    
    abstract val dataClass: Class<out StudyObject>
    
    override var id: Long
        get() = data.id
        set(id) {
            data.id = id
        }
    
    override val position: Int
        get() = data.position
    
    override val name: String
        get() = data.title
    
    override fun getPrevChild(current: StudyNode?): StudyNode? {
        return children.lastOrNull {
            !it.wasDeleted && (current == null || it.position < current.position)
        }
    }
    
    override fun getNextChild(current: StudyNode?): StudyNode? {
        return children.firstOrNull {
            !it.wasDeleted && (current == null || it.position > current.position)
        }
    }
    
    override fun getChildById(id: Long) = mapNodes[id]
    
    override fun getChildByClassAndId(clazz: Class<out StudyNode>, id: Long): StudyNode? {
        return if (childDataClass == clazz) {
            getChildById(id)
        } else {
            children.mapNotNull { it.getChildByClassAndId(clazz, id) }
                    .firstOrNull()
        }
    }
    
    override fun getChildByPosition(position: Int): StudyNode? {
        return children.first { it.position == position }
    }
    
    protected open fun getChildDataList(stepikApiClient: StepikApiClient): List<StudyObject> {
        return emptyList()
    }
    
    final override fun init(project: Project?, stepikApiClient: StepikApiClient?, parent: StudyNode?) {
        beforeInit()
        
        this.parent = parent
        this.project = project
        
        if (stepikApiClient != null) {
            val processed = mutableListOf<StudyNode>()
            val children = this.children.toMutableList()
            
            for (data in getChildDataList(stepikApiClient)) {
                var child = mapNodes[data.id]
                if (child == null) {
                    try {
                        child = childClass.newInstance()
                        children.add(child)
                    } catch (e: Exception) {
                        logger.warn("Can't get new instance for child", e)
                        break
                    }
                }
                child!!.data = data
                child.wasDeleted = wasDeleted
                processed.add(child)
            }
            _children = children.sortedWith(StudyNodeComparator)
            
            children.filterNot { it in processed }
                    .forEach { it.wasDeleted = true }
        }
        
        children.forEach { it.init(project, stepikApiClient, this) }
    }
    
    open fun beforeInit() {
    
    }
    
    override fun reloadData(project: Project, stepikApiClient: StepikApiClient) {
        if (loadData(stepikApiClient, data.id)) {
            init(project, stepikApiClient)
        }
    }
    
    protected abstract fun loadData(stepikApiClient: StepikApiClient, id: Long): Boolean
    
    override fun resetStatus() {
        children.forEach { it.resetStatus() }
        status = UNCHECKED
    }
    
    override fun passed() {
        status = SOLVED
    }
    
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        
        val node = other as Node
        
        if (wasDeleted != node.wasDeleted) return false
        if (data != node.data) return false
        
        if (children != node.children) return false
        
        return status === node.status
    }
    
    override fun hashCode(): Int {
        var result = data.hashCode()
        result = 31 * result + children.hashCode()
        result = 31 * result + wasDeleted.hashCode()
        result = 31 * result + status.hashCode()
        return result
    }
    
    companion object {
        private val executor = Executors.newFixedThreadPool(5)
    }
}
